airflow/stats.py

```diff
@@ -72,20 +73,71 @@ def timer(cls, *args, **kwargs) -> TimerProtocol:
         """Timer metric that can be cancelled"""
 
 
-class DummyTimer:
-    """No-op timer"""
+class Timer:
+    """
+    Timer that records duration, and optional sends to statsd backend.
+
+    This class lets us have an accurate timer with the logic in one place (so
+    that we don't use datetime math for duration -- it is error prone).
+
+    Example usage:
+
+    .. code-block:: python
+
+        with Stats.timer() as t:
+            # Something to time
+            frob_the_foos()
+
+        log.info("Frobbing the foos took %.2f", t.duration)
+
+    To send a metric:
+
+    .. code-block:: python
+
+        with Stats.timer("foos.frob"):
+            # Something to time
+            frob_the_foos()
+
+
+    Or both:
+
+    .. code-block:: python
+
+        with Stats.timer("foos.frob") as t:
+            # Something to time
+            frob_the_foos()
+
+        log.info("Frobbing the foos took %.2f", t.duration)
+
+    """
+
+    # pystatsd and dogstatsd both have a timer class, but present different API
+    # so we can't use this as a mixin on those, instead this class is contains the "real" timer
+
+    _start_time: Optional[int]
+    duration: Optional[int]
+    ms: Optional[int]
+    _send: bool = False
```

Comment:
How is this `_send` attribute behaving here? Or only for compatibility with `Timer` in `pystatsd `?

Reply:
And the `ms` attribute above.

May be dumb question ;-)

Reply:
Oh yes, these aren't needed anymore. I was planning on making this a subclass of the timer, bit pystatsd and dogstatsd are too different to work. So I went with encapsulation instead.

These can be removed

Reply:
Done.